### ISSUE TITLE
[BUG] Can't upload large image files via web UI

### ISSUE THREAD

**Issue by noahgildersleeve** [harvester/harvester#1904]
**Describe the bug**

When you try to upload large ISO or image files it errors out
**To Reproduce**
1. Navigate to the images page
2. Select upload file
3. Upload a file that is larger than 2 GB. I tested it with a server2019 ISO

**Expected behavior**

The upload should complete successfully. 
**Support bundle**

[supportbundle_c2f07121-0b02-46ef-8114-d5867098fd10_2022-02-09T19-34-58Z.zip](https://github.com/harvester/harvester/files/8035677/supportbundle_c2f07121-0b02-46ef-8114-d5867098fd10_2022-02-09T19-34-58Z.zip)

**Environment:**
 - Harvester ISO version: master-f303b27e-head
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM virtualized environment

**Additional context**
This doesn't happen on `v1.0.0`. It was working fairly recently and stopped working in the last few days while I was testing. It appears to start uploading then resets the connection with a 500 error.
![Screenshot from 2022-02-09 11-33-09](https://user-images.githubusercontent.com/83787952/153277005-ed0c00da-8290-48e3-ad7d-58448f717019.png)

**Comment by w13915984028** [comment]
A quick test:

image harvester-master-amd64.iso is downloaded independently (url: https://releases.rancher.com/harvester/master/harvester-master-amd64.iso)

harvester-master-amd64-2.iso, harvester-master-amd64-3.iso are downloaded in parallel,

all are successful


![image](https://user-images.githubusercontent.com/31133476/155303889-485af7dd-b48e-40d2-b2c2-f96cac3786a6.png)

**Comment by w13915984028** [comment]
@noahgildersleeve could you please try it again in your environment? and, when it happened, did you check your setup, was there adequate storage size ? thanks.

**Comment by w13915984028** [comment]
Update:

Reproduced with `"uploading large image"`, after that:

Error logs in Harvester POD
```
time="2022-02-23T12:37:53Z" level=error msg="Error during subscribe io: read/write on closed pipe"
time="2022-02-23T12:37:54Z" level=error msg="Error writing ping" error="write tcp 10.52.0.212:49954->10.53.156.141:443: i/o timeout"
time="2022-02-23T12:37:59Z" level=error msg="Error writing ping" error="write tcp 10.52.0.212:49954->10.53.156.141:443: i/o timeout"
time="2022-02-23T12:38:04Z" level=error msg="Error writing ping" error="write tcp 10.52.0.212:49954->10.53.156.141:443: i/o timeout"
time="2022-02-23T12:38:09Z" level=error msg="Error writing ping" error="write tcp 10.52.0.212:49954->10.53.156.141:443: i/o timeout"
```

![image](https://user-images.githubusercontent.com/31133476/155324225-fdeba8af-d13c-4e59-96d3-28928d3235c1.png)

**Comment by w13915984028** [comment]
When uploading big size image (e.g. > 1G), the issue is highly reproduceable.

After the issue happened, refreshing `Harvester WebUI`  may encounter blank page with `"Handler disconnected"`.

The **workaround** is using below command to delete the Harvester pod, then the WebUI will be reusable.
`kubectl delete pod harvester-pod-name -n harvester-system`

`harvester-pod-name` is like:
```
kubectl get pods -n harvester-system
NAME                                                   READY   STATUS    RESTARTS   AGE
harvester-d544ddb6f-phwhd                              1/1     Running   0          15h
```

**Comment by w13915984028** [comment]
Another test:
`When uploading speed is fast, for instance, the client PC and Harvester cluster  are in same network, the harvester POD had such error message right after the uploading action starts.`

but the uploading is still on-going

kubectl logs harvester-d544ddb6f-kbp2s -n harvester-system
```
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4224768"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4241125"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4257536"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4273875"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4306625"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4323072"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4339375"
time="2022-02-24T21:59:11Z" level=error msg="remotedialer buffer exceeded, length: 4372143"
...
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 11774178"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 9922849"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 7760161"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 7776446"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 6318369"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 6334636"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 5793955"
time="2022-02-24T22:11:18Z" level=error msg="remotedialer buffer exceeded, length: 4335806"
```

![image](https://user-images.githubusercontent.com/31133476/155617046-53b7c55e-8c47-4d8a-9e57-c1dd0d34a878.png)

3.7 GB ISO is uploaded successfully in 10 minutes
![image](https://user-images.githubusercontent.com/31133476/155617360-84d5db74-9a1e-4c9e-b017-c67f2abb9ab8.png)

**Comment by w13915984028** [comment]
`3 Harvester master-head (today) ISO are uploaded in parallel, all are successful`

even though Harvester POD had error message, it looks self-healed.

 kubectl logs harvester-d544ddb6f-kbp2s -n harvester-system
```
time="2022-02-24T22:23:57Z" level=error msg="remotedialer buffer exceeded, length: 4225524"
time="2022-02-24T22:23:57Z" level=error msg="remotedialer buffer exceeded, length: 4258292"
time="2022-02-24T22:23:57Z" level=error msg="remotedialer buffer exceeded, length: 4273875"
time="2022-02-24T22:23:57Z" level=error msg="remotedialer buffer exceeded, length: 4306643"
time="2022-02-24T22:23:57Z" level=error msg="remotedialer buffer exceeded, length: 4323828"

...
time="2022-02-24T22:26:09Z" level=error msg="remotedialer buffer exceeded, length: 10143932"
time="2022-02-24T22:26:09Z" level=error msg="remotedialer buffer exceeded, length: 8243334"
time="2022-02-24T22:26:09Z" level=error msg="remotedialer buffer exceeded, length: 6146182"
time="2022-02-24T22:26:09Z" level=error msg="remotedialer buffer exceeded, length: 6146236"
time="2022-02-24T22:26:09Z" level=error msg="remotedialer buffer exceeded, length: 6178950"
W0224 22:26:35.672889       8 transport.go:260] Unable to cancel request for *client.addQuery
W0224 22:26:35.673150       8 transport.go:260] Unable to cancel request for *client.addQuery
```

![image](https://user-images.githubusercontent.com/31133476/155618418-cca12c95-ab2c-43ff-9e52-6bd17c24d576.png)

![image](https://user-images.githubusercontent.com/31133476/155619423-76ded792-4f50-45af-9466-550ff61a881a.png)

![image](https://user-images.githubusercontent.com/31133476/155619795-3e8ab148-9e0e-4dbf-a9fd-8e66c9cb75f5.png)

**Comment by noahgildersleeve** [comment]
Tested in `master-bfd6a8a9-head`
## Verification steps
1. Deploy Harvester from master-head ISO
2. Upload large ISO
3. Wait for it to complete
4. Upload second large ISO
5. Wait for it to complete

The images are uploading, but after the first ISO I did get the `"Handler disconnected"` error. I'm including the support bundle, but I'm going to create a new bug for that issue since it's out of scope of the original report.
![Screenshot from 2022-02-24 15-08-14](https://user-images.githubusercontent.com/83787952/155623020-667aa7ee-0b00-4421-9684-ccebf194b6ff.png)
![Screenshot from 2022-02-24 14-54-19](https://user-images.githubusercontent.com/83787952/155623022-bb28dbd3-33d4-4b3c-96b4-d8efe12230fa.png)
[supportbundle_f3da3adc-2aef-401f-beaa-bb13ac005b1c_2022-02-24T22-59-06Z.zip](https://github.com/harvester/harvester/files/8137140/supportbundle_f3da3adc-2aef-401f-beaa-bb13ac005b1c_2022-02-24T22-59-06Z.zip)

**Comment by w13915984028** [comment]
As per the support bundle in issue description, following observation deserves a further discussion.

https://github.com/harvester/harvester/issues/1904#issue-1128947127

https://github.com/harvester/harvester/files/8035677/supportbundle_c2f07121-0b02-46ef-8114-d5867098fd10_2022-02-09T19-34-58Z.zip

**Comment by w13915984028** [comment]
Harvester node running environment:
**12 core CPU with 16G memroy**

**OOM is in kernel log, Harvester (POD) is killed by kernel**

-- Logs begin at Wed 2022-02-09 18:12:09 UTC, end at Wed 2022-02-09 19:35:17 UTC. --
Feb 09 18:12:09 localhost kernel: Linux version 5.3.18-59.37-default (geeko@buildhost) (gcc version 7.5.0 (SUSE Linux)) #1 SMP Mon Nov 22 12:29:04 UTC 2021 (d10168e)
Feb 09 18:12:09 localhost kernel: Command line: BOOT_IMAGE=(loop0)/boot/vmlinuz console=tty1 root=LABEL=COS_ACTIVE cos-img/filename=/cOS/active.img panic=0 net.ifnames=1 rd.cos.oemtimeout=120 rd.cos.oemlabel=COS_OEM

`18:12:09 localhost kernel: Memory: 2065184K/16776676K available (12291K kernel code, 2236K rwdata, 4592K rodata, 2636K init, 13908K bss, 452780K reserved, 0K cma-reserved)`

`Feb 09 18:12:09 localhost kernel: smpboot: Total of 12 processors activated (91199.95 BogoMIPS)`


-----
`Feb 09 19:32:34 harvester-master kernel: longhorn invoked oom-killer: gfp_mask=0x100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=999`
Feb 09 19:32:34 harvester-master kernel: CPU: 9 PID: 7737 Comm: longhorn Tainted: G        W           5.3.18-59.37-default #1 SLE15-SP3
Feb 09 19:32:34 harvester-master kernel: Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.13.0-1ubuntu1.1 04/01/2014



Feb 09 19:32:34 harvester-master kernel: Node 0 Normal free:69,264kB min:110,776kB low:125,668kB high:140,560kB active_anon:13,456,596kB inactive_anon:13,968kB active_file:7372kB inactive_file:1012kB unevictable:12760kB writepending:0kB present:14680064kB managed:14389784kB mlocked:12760kB kernel_stack:45992kB pagetables:57984kB bounce:0kB free_pcp:13336kB local_pcp:1360kB free_cma:0kB

Feb 09 19:32:34 harvester-master kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=6066f28a301347d9da0892fbaa8b71148d117ceba506c8fb5a47757aca0c12fd,mems_allowed=0,global_oom,task_memcg=/kubepods/burstable/pod623e2cd3-bedc-4ba7-a145-4eee01e6055c/0e94be5f668d12a7312e6b59fbc2925cf195cdaf70a7f802106f14e38b5f3bea,task=harvester,pid=1513,uid=0

`Feb 09 19:32:34 harvester-master kernel: Out of memory: Killed process 1513 (harvester) total-vm:15594216kB, anon-rss:8411872kB, file-rss:0kB, shmem-rss:0kB`

Feb 09 19:32:34 harvester-master kernel: oom_reaper: reaped process 1513 (harvester), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB

**Comment by w13915984028** [comment]
Timeline of what happened.

`The Harvester POD is killed when uploading was ongoing`, that's why the WebUI shows 500 error, and the uploading failed. 

![image](https://user-images.githubusercontent.com/31133476/156389145-c12996b1-0729-43e8-88f9-46e331912ffe.png)


![image](https://user-images.githubusercontent.com/31133476/156376382-0fd7de29-878e-4ba2-9181-ab79a17b89bb.png)

**Comment by w13915984028** [comment]
Why OOM?

When networking speed is fast enough (e.g, Harvester and uploading PC are in same/near network),  the networking transport is faster than disk writing, the data will be cached in Harvester POD, that is in memory.  Especially when multi uploading are in parallel.



In my local test, I encountered following message in Harvester POD, it is when Harvester saving data :

```
time="2022-02-24T14:06:33Z" level=error msg="remotedialer buffer exceeded, length: 4257680"
time="2022-02-24T14:06:33Z" level=error msg="remotedialer buffer exceeded, length: 4273155"


time="2022-02-24T14:06:43Z" level=error msg="remotedialer buffer exceeded, length: 7747427"
time="2022-02-24T14:06:43Z" level=error msg="remotedialer buffer exceeded, length: 7761030"
time="2022-02-24T14:06:43Z" level=error msg="remotedialer buffer exceeded, length: 7761246"
time="2022-02-24T14:06:43Z" level=debug msg="Wrote ping"
time="2022-02-24T14:06:44Z" level=error msg="remotedialer buffer exceeded, length: 7768236"
```


![image](https://user-images.githubusercontent.com/31133476/156377209-8bc514ea-0260-4a99-b591-245e51b34dab.png)

**Comment by w13915984028** [comment]
Considering system requirements:

https://docs.harvesterhci.io/v1.0/

![image](https://user-images.githubusercontent.com/31133476/156378633-d107ae9c-3eac-4d56-8bcc-8744441450fa.png)

**Comment by w13915984028** [comment]
Summary:

To reproduce 1904 issue:
1.  Give Harvester node fewer memory, say 16G
2. Has fast networking, the uploading speed is faster than LH image writing
3. Uploading a couple of BIG ISO images in parallel
as long as Harvester POD caches enough uploading data, it is highly possible to reproduce.

Why my test https://github.com/harvester/harvester/issues/1904#issuecomment-1050335273 does not reproduce it? because my WIFI network speed is not fast enough, 3.7G ISO image takes 16 min, in @noahgildersleeve test, 5.17G files takes 12 min or less.

@noahgildersleeve : Could you please re-test with at lest 32G memory for the Harvester node.

@guangbochen @yasker : 
For uploading image, should we limit the parallel count ? e.g, at most 2 uploading in parallel.

**Comment by w13915984028** [comment]
BTW, the feature here will also help:
https://github.com/harvester/harvester/issues/1154
[FEATURE] Hardware Spec preflight checking in the installer

**Comment by schmitch** [comment]
well the thing is, 32gb is a ton of memory. not even vsphere has that amount. esxi only needs 8gig while vcenter needs 10g

**Comment by noahgildersleeve** [comment]
@w13915984028 I retested this with 32GB of memory on my VM. It seemed to work fine. I uploaded both large image files one right after the other and the frontend didn't fail. I attached the support bundle and screenshots. Does this mean that this ticket is OK to close? It seems like the original scope of this ticket isn't the same as what the issue ended up being. 
![Screenshot from 2022-03-02 10-50-50](https://user-images.githubusercontent.com/83787952/156428848-8a272291-15e5-420c-8055-7b81e3307bba.png)
![Screenshot from 2022-03-02 10-50-43](https://user-images.githubusercontent.com/83787952/156428849-8bb098c2-596d-4396-b39a-863da627cd25.png)
![Screenshot from 2022-03-02 10-49-42](https://user-images.githubusercontent.com/83787952/156428851-83991039-9673-44c5-8cc7-f792a9915454.png)
[supportbundle_39ed8db0-18c8-4e1b-914b-cada6a5645a3_2022-03-02T18-52-30Z.zip](https://github.com/harvester/harvester/files/8172188/supportbundle_39ed8db0-18c8-4e1b-914b-cada6a5645a3_2022-03-02T18-52-30Z.zip)

**Comment by w13915984028** [comment]
Thanks @noahgildersleeve , `the new log proves our analysis`.  In your test, the uploading speed is fairly high

I checked the support bundle, the `Harvester POD` has `59,162 lines of log`, `only lasts 10 seconds, from 18:50:16 to 18:50:26,  all previous logs are overwritten)`

Here we see, the maximum buffer exceed is `1,528,228,071`,  apparently, the peak value could be higher up to 5G. But as the VM has 32G memory, it is enough to accommodate, OOM did not happen.  But this situation is still very risky.

`The enhancement #1956, is essential.`

@yasker @guangbochen 


2022-03-02T18:50:16.789512320Z time="2022-03-02T18:50:16Z" level=error msg="remotedialer buffer exceeded, length: 1,528,228,071"

![image](https://user-images.githubusercontent.com/31133476/156460719-7ef739c0-9293-440c-85eb-1ddba628505c.png)

![image](https://user-images.githubusercontent.com/31133476/156460750-a36537f7-30fd-4734-b5b7-2ba3855ab62d.png)

![image](https://user-images.githubusercontent.com/31133476/156460831-96ff8c17-6f26-4500-9820-501814a63813.png)


```

2022-03-02T18:50:16.789512320Z time="2022-03-02T18:50:16Z" level=error msg="remotedialer buffer exceeded, length: 1528228071"
2022-03-02T18:50:16.789555851Z time="2022-03-02T18:50:16Z" level=error msg="remotedialer buffer exceeded, length: 1528228071"
2022-03-02T18:50:16.789559511Z time="2022-03-02T18:50:16Z" level=error msg="remotedialer buffer exceeded, length: 1526753511"
2022-03-02T18:50:16.789562361Z time="2022-03-02T18:50:16Z" level=error msg="remotedialer buffer exceeded, length: 1526753511"

2022-03-02T18:50:18.270682084Z time="2022-03-02T18:50:18Z" level=error msg="remotedialer buffer exceeded, length: 1284740734"
2022-03-02T18:50:18.270917467Z time="2022-03-02T18:50:18Z" level=error msg="remotedialer buffer exceeded, length: 1284544126"
2022-03-02T18:50:18.271053968Z time="2022-03-02T18:50:18Z" level=error msg="remotedialer buffer exceeded, length: 1284544126"

2022-03-02T18:50:26.547697487Z time="2022-03-02T18:50:26Z" level=error msg="remotedialer buffer exceeded, length: 4599177"
2022-03-02T18:50:26.547841129Z time="2022-03-02T18:50:26Z" level=error msg="remotedialer buffer exceeded, length: 4566409"
2022-03-02T18:50:26.547962960Z time="2022-03-02T18:50:26Z" level=error msg="remotedialer buffer exceeded, length: 4533587"
```

**Comment by w13915984028** [comment]
@noahgildersleeve 
I agree with you that this issue could be closed, all potential enhancements are tracked with new issues, they have connections with 1904.
#1904 gives us a good chance to investigate details and robustness of Harvester, thanks.

**Comment by yasker** [comment]
@w13915984028 @noahgildersleeve I feel we should keep this issue since this is what does the user see in the end. Whatever fix in #1955 or #1956 should be considered as a part of this issue. We're narrowing down this issue but haven't fixed it yet. I would prefer us to close it only when the user won't be impacted by the issue. cc @rebeccazzzz

**Comment by w13915984028** [comment]
Good to keep this root issue (#1904) open.

#1956 looks to be the key, and #1940 can also get benefits from #1956

**Comment by w13915984028** [comment]
The latest Harvester version with bumped Rancher version, which has pause/resume protocol, both #1904 and #1956 are solved.

Uploading big ISO images will not cause Harvester POD exhausting large amount of memory.

master-d8a7dfae-head
![image](https://user-images.githubusercontent.com/31133476/159266876-6615cb65-d9f9-4cd7-bb56-6cc165262df3.png)
![image](https://user-images.githubusercontent.com/31133476/159267053-5483fd29-56d9-4c90-b010-6f411ece2b8c.png)

**Comment by noahgildersleeve** [comment]
Tested in Harvester-Tested in master-d8a7dfae-head.
## Verification steps
1. Upload two Windows Server ISOs (>5GB) to Harvester in KVM virtualized environment to simulate very fast network
2. Check harvester k8s pd logs
3. Check memory usage

Memory usage seemed fine and there weren't too many remotedialer errors
```
time="2022-03-21T21:36:42Z" level=error msg="Error during subscribe websocket: close sent"
time="2022-03-21T21:36:54Z" level=error msg="Error during subscribe websocket: close sent"
time="2022-03-21T21:37:44Z" level=error msg="Error during subscribe websocket: close sent"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4227072"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4259840"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4292608"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4259840"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4292608"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4325376"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4358144"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4390912"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4390912"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4423680"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4456448"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4456448"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4489216"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4521984"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4554752"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4587520"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4566199"
time="2022-03-21T21:38:55Z" level=error msg="remotedialer buffer exceeded, length: 4582574"
time="2022-03-21T21:42:20Z" level=error msg="Error during subscribe websocket: close sent"
```
![Screenshot from 2022-03-21 14-42-38](https://user-images.githubusercontent.com/83787952/159368769-e748e129-e4f7-4af6-a072-ba42cbb804be.png)
